sdk/src/main/scala/com/horizen/account/receipt/LogsBloom.scala

```diff
@@ -0,0 +1,105 @@
+package com.horizen.account.receipt
+
+import scorex.crypto.hash.Keccak256
+import com.horizen.account.receipt.LogsBloom.BLOOM_FILTER_LENGTH
+import com.horizen.evm.interop.EvmLog
+import com.horizen.utils.BytesUtils
+
+class LogsBloom() {
+  private var bloomFilter = Array.fill[Byte](BLOOM_FILTER_LENGTH)(0)
+
+  def this(bloomFilter: Array[Byte]) = {
+    this()
+
+    require(bloomFilter.length == BLOOM_FILTER_LENGTH)
+    this.bloomFilter = bloomFilter
+  }
+
+  def addLogToBloomFilter(log: EvmLog): Unit = {
+    addBytesToBloomFilter(log.address.toBytes)
+    log.topics.foreach(topic => addBytesToBloomFilter(topic.toBytes))
+  }
+
+  def addBytesToBloomFilter(
+      data: Array[Byte]
+  ): Unit = {
+    val (bloomFilterIndexes, bloomFilterValues) = getBloomFilterValues(data)
+
+    bloomFilterIndexes.zipWithIndex
+      .foreach({ case (bloomFilterIndex, i) =>
+        bloomFilter(bloomFilterIndex) =
+          (bloomFilter(bloomFilterIndex) | bloomFilterValues(i)).toByte
+      })
+  }
+
+  def setBytes(data: Array[Byte]): Unit = {
+    require(bloomFilter.length == BLOOM_FILTER_LENGTH)
+
+    bloomFilter.zipWithIndex
+      .foreach({ case (bloomByte, i) =>
+        bloomFilter(i) = (bloomByte | data(i)).toByte
+      })
+  }
+
+  def contains(data: Array[Byte]): Boolean = {
+    val (dataIndexes, dataValues) = getBloomFilterValues(data)
+
+    dataIndexes.zipWithIndex.foreach({ case (dataIndex, i) =>
+      if (bloomFilter(dataIndex).&(dataValues(i)) != dataValues(i)) {
+        return false
+      }
+    })
+
+    true
+  }
+
+  private def getBloomFilterValues(
+      data: Array[Byte]
+  ): (Array[Int], Array[Int]) = {
+    val hashBuffer = Keccak256.hash(data)
+
+    val bloomFilterValues: Array[Int] =
+      Array(1, 3, 5).map(e => 1 << (hashBuffer(e) & 0x7))
+
+    val bloomFilterIndexes: Array[Int] = Array((0, 1), (2, 3), (4, 5)).map(e =>
```

Comment:
Personally, I'm not a fan of using these interator-map-constructs for something that will only ever be executed exactly three times, It doesn't improve readability. Not sure if there is any performance difference, so I guess it comes down to personal preference :wink:

Reply:
In my opinion the first iterator looks cleaner, second one is kinda more confusing because of the tuples. I actually think that there is performance degradation regarding the second one because the array needs to be created just to retrieve the indexes of hashBuffer that need to be set. Second one was created just to continue the coding style based on the first one. I'll revert the changes to make the code more readable and optimized 😄